### DISCUSSION TITLE
useRouter returns wrong query object when page is cached.

### DISCUSSION THREAD

**Discussion by kartikag01** [vercel/next.js#21179] [Help]
**What version of Next.js are you using?**

10.0.3

**What version of Node.js are you using?**

10.16.3

**What browser are you using?**

Chrome

**What operating system are you using?**

Ubuntu

**How are you deploying your application?**

AWS Cloudfront 

**Describe the Bug**

`query` object from `useRouter();` have old values when page is served from cache, both before and after hydration. 

**Expected Behavior**

`useRouter();` should returns correct query object after hydration process,

**To Reproduce**

This problem start appearing when we changed `Query String Forwarding and Caching` rules in cloudfront from `Forward all, cache based on all` to `Forward all, cache based on whitelist`

I have a simple page, just to print query object. 

```
// pages/test-page.js

function TestPage() {
  const router = useRouter();
  return (
    <div>
      <h3>Query</h3>
      <h5>{JSON.stringify(router.query)}</h5>
    </div>
  );
}
```
### First time URL visit
https://domain.com/test-page?tech=react&year=2013

output is straightforward
![image](https://user-images.githubusercontent.com/10574227/104341786-2f8cc080-5520-11eb-8598-8a45fe4fd549.png)

Since this is first time we visited this page, so it came from source not from the cache.

### Second time URL visit
now we have different values in query
https://domain.com/test-page?tech=nextjs&year=2016

Now since page was cached, it should give the same response but after next.js hydration process it should print the correct values.  

But even though after hydration process page given the old values.
![image](https://user-images.githubusercontent.com/10574227/104341717-1d128700-5520-11eb-9657-c5bd3d13899a.png)

Query String whitelist doesn't contains `tech` and `year` as we don't want to cache pages on these query params, but on client side we should to have its correct values, regardless with what values page was opened first time.

**Comment by thescientist13** [comment]
Yeah, experiencing this challenge as well.  Would be great if there was a "no hydrate" option for `useRouter` hook to always force it to read from the URL in the browser.

**Reply by weee-shinan** [reply]
Is there any other way to resolve this?

**Comment by weee-shinan** [comment]
@kartikag01 Is there any other way to resolve this?
